apply suggested improvements from app theme DX part2 (#96)
vaadin/docs: articles/theming/application-theme.asciidoc

Application Theme is the standard way to organize themes in a Vaadin application, making it easier to switch between themes and allow packaging a theme to reuse in other Vaadin applications.
--

== Limitations
--
The followings options are not currently available:
--

> <1> Activating multiple application themes at the same time.

Author: The idea was from one DX tester who wanted to use a set of styles from one theme, and another set of styles from another theme. So he wanted to add 2 theme jars to the project and activating them by something like: 
```
@Theme("theme-A", "theme-B")
```
or 
```
@Theme("theme-A")
@Theme("theme-B")
public class Something { }
```
And OFC the API and a runtime exception would inform him those possibilities are not in the picture, but it would be much nicer to have them directly in docs.

> <1> Activating multiple application themes at the same time.

Reviewer: Then this should not be under the `not currently available` as as such it will never be implemented.

> <1> Activating multiple application themes at the same time.

Reviewer 2: We should remove the “currently”. If we have plans on addressing some of these limitations in the future, add a link to a relevant issue in GitHub instead.

> <1> Activating multiple application themes at the same time.

Author: Done.

Changed the title completely from `not currently available` to `not available`.

<2> Switching between application themes at runtime.
<3> Extending from multiple parent themes.

== Define Application Themes

The following describes a sample file system structure for "my-theme":

.Figure 1: A sample application theme called "my-theme"
[source, filesystem]
----
frontend/
└── themes/ <1>
    └── my-theme/ <2>
        ├── components <3>
        │   ├── vaadin-button.css <4>
        │   └── vaadin-text-field.css <4>
        ├── some-images <5>

>         ├── some-fonts <5>

Author: I think the explanation below the figure is clear enough to show they are optional. Also, in another part of this doc, there is an example of using relative path used in CSS which refers to images inside "some-images". Although that example is not update from "images" to "some-images".

>         ├── some-fonts <5>

Reviewer: Could just be some-folder and the explanation could mention that it can be used  for structuring the theme files.
As now it sort of leaves out the chance of having css inside folders.

>         ├── some-fonts <5>

Author: Done. 

I kept `/some-images` there since it's been referenced in another example. But changed the `/some-fonts` to `some-folders` and changed the description to a more general statement to include all sorts of files.

<6> CSS that is always applied to the root of the page (for <<Theming Embedded Web Components,embedded web component>>). It should be named as `document.css`.
<7> Automatically loaded master stylesheet. Other non-component-specific stylesheets need to be imported through this file. It should be named as `styles.css`.
<8> Theme configurations (e.g. assets from `node_modules` or <<Extending an existing application theme,parent theme>> properties) are defined in this file. It should name as `theme.json`.

.Create the theme folder manually
[NOTE]
The theme folder will not be autogenerated and must be created by the developer.

=== Application Theme Folder
Application themes should be placed under the `frontend/themes/` folder of a Vaadin project's root location.

The `themes` folder location is not configurable.

Each theme should be created in a separate folder, the folder name is also the theme name that is used for `@Theme`.

For example:

* `@Theme("alpaca")`: the files should be in `frontend/themes/alpaca/`
* `@Theme("frost-bite")`: the files should be `in frontend/themes/frost-bite/`

=== Global Styles

> Application-wide styles which are used across multiple views shall be defined in `styles.css`. Since this is an automatically loaded CSS file it should be named as `styles.css`, located directly under theme folder's root.

Author: Done.

=== Component Styles

> Component-specific style sheets should be placed under the `components` folder.

Author: This is something suggested by grammarly to change the passive tone of the sentence. ;-) 
It suggests to convert "are placed in ..." into "should be placed in ... by [developer]" and "[developer]" is omitted everywhere since it is obvious.

}
----

Now, assuming the theme folder named as "my-theme", we can refer to the assets by adding `themes/my-theme/` to the local URIs defined in `theme.json`:

In stylesheet:

[source, CSS]
----
background-image: url('themes/my-theme/fortawesome/icons/snowflake.svg');
----

In java code:

[source, java]
----
Image snowImg = new Image("themes/my-theme/fortawesome/icons/snowflake.svg", "snow");
----

> NOTE: `assets` block is not accepting array of mappings. For defining each mapping, a separate `assets` block should be defined.

Author: Nice one! I added this to the docs as separate example.

----
include::src/main/java/com/vaadin/flow/tutorial/theme/UsingComponentThemes.java[tags=lumo-dark, indent=0, group=Java]
----
See <<theme-variants#global-variants,Theme Variants>>.

NOTE: Any changes to theme folder contents (add/remove files, file modifications) are automatically picked up by webpack.
If a running app has live reload available, you can wait for it to reload the changes, otherwise, you can manually refresh the browser.

NOTE: If you notice the following kind of webpack error after removing any files/folders from theme folder `Error: ENOENT: no such file or directory, open '.../themes/my-theme/styles.css'`, please manually refresh the browser.

== Extending an existing application theme

The application theme can extend an existing application theme, usually called **parent theme**, that is either located in
the project themes folder or as a <<Creating a Reusable Application Theme,reusable>> theme dependency jar.

Project themes folder places are `frontend/themes`, `src/main/resources/static` and `src/main/resources/META-INF/resources/themes`.

> Process of extending from a parent theme is similar to having a theme folder except for the fact of adding "parent" key to the `theme.json`, so:

Author: That sentence supposed to say the same thing which obviously doesn't! I rephrased it!

. Create a theme folder in your application with a different name from the parent theme's name, under `frontend/themes/`.
. Use the new theme folder's name in `@Theme("...")`.

> . In the `theme.json` add the parent theme's name, with the key `parent`:

Author: Done.

[source, JSON]
----
{
  "parent": "acme-base"
}
----

> Now besides to overriding parent theme's styles, new styles can be added also through the local theme folder and its configurations.

Reviewer: ```suggestion
The extending theme can override parent theme styles as well as add its own.
```

== Creating a Reusable Application Theme

Application theme can be packaged and reused in other applications.
A reusable theme is a jar file containing only the theme files, and an optional java class (e.g. `Dependency.java`) with `NpmPackage`
declarations for npm packages required by the theme.

=== Creating the theme jar

The easiest way to create a reusable theme jar is to create a maven project with the following structure:

.Figure 2: Folder structure of a maven project needed for creating a theme jar
[source, filesystem]
----
[project-root]
├── src
│   └── main
│       ├── java
│       │   └── com

> │       │       └── example

Reviewer: com.example is bad from the spring perspective.
com.vaadin.flow.component would be better

Author: Just wanted to keep the path shorter in the Figure. Also wanted to avoid "vaadin" to prevent any possible confusion about name of this package, since someone may infer it should have some "vaadin" inside it or to be exactly that `com.vaadin.flow.component`. 

Changed it to `org.brandname.theme`

│       │           └── Dependencies.java (optional)
│       └── resources
│           └── META-INF
│               └── resources
│                   └── themes
│                       └── my-theme
│                           ├── components/
│                           ├── ...
│                           ├── ...
│                           ├── styles.css
│                           └── theme.json
└── pom.xml
----

The detailed steps are as follows:

. Create a new maven project using either IDE or Maven. Example maven command:
`mvn archetype:generate -DgroupId=com.example -DartifactId=reusable -DarchetypeArtifactId=maven-archetype-quickstart -DarchetypeVersion=1.4 -DinteractiveMode=false`

> . Remove `src/test` as well as `App.java` from `src/main/java/`. (Optional)

Reviewer: As the App.java is not used it shouldn't be added to the generated jar and thus should be removed.

Author: Since not everybody may create the project using that maven archetype, thus there is no `App.java` to be removed. I think changing `Optional` to `Recommended` would make a clear point, since it is not actually that mandatory and having `App.java` would not harm anything.

}
----

The extending theme can override parent theme styles as well as add its own.

== Creating a Reusable Application Theme

Application theme can be packaged and reused in other applications.
A reusable theme is a jar file containing only the theme files, and an optional java class (e.g. `Dependency.java`) with `NpmPackage`
declarations for npm packages required by the theme.

=== Creating the theme jar

> The easiest way to create a reusable theme jar is to create a maven project with the following structure:

Reviewer: This is a bit missleading as it's not the easiest way it's the way with maven.
```suggestion
To create a reusable theme jar with maven, create a project with the following structure:
```

> . Create a new maven project using either IDE or Maven. Example maven command:

Reviewer: I wouldn't merge create with IDE and create with maven archetype as the steps are for using archetype where as the create with IDE is IDE specific and should just say to create the shown structure. So there should be 2 paths to follow until `Create the folders ...` point.

Author: I think it is better not to go into details about **how to create an empty maven project** since this document is not about helping users to find out how to use/create maven projects. I would also exclude the second step (removing the App.java file) since that file's existence is dependant on how users are creating that maven project. To wrap up, I would merge/modify first to steps: like this:
```
1. Create a new empty maven project (using either your preferred IDE's wizard or through Maven CLI) and 
make sure no unused java classes or resources exist in the project, then continue with the next steps. 
```

Reviewer: I would in this instance remove all the steps and just show the contents of the pom. as everything else is in the figure.
Also the header `=== Creating the theme jar` should be renamed to something else.. perhaps `=== Reusable theme project structure` or `=== Reusable theme contents`

Author: Done.

> . Create the folders `src/main/resources/META-INF/resources/themes/` and the reusable theme folder.

Reviewer: ```suggestion
. Create the folders `src/main/resources/META-INF/resources/themes/[reusable-theme]`.
```

> . Add the theme's assets and resources to a new theme folder under `themes/`.

Reviewer: ```suggestion
. Add the theme files to the reusable theme folder.
```
As you created it just above you shouldn't create a new theme folder here.
